### DISCUSSION TITLE
From XML to Postgres...in docker...

### DISCUSSION THREAD

**Discussion by cappelaere** [opendcs/opendcs#146] [Q&A] [answered]
I can deploy opendcs in a container, connect to wcda and get messages in the default XML database.
I can create an AWS RDS (postgresql) and run the createDB.sh script to load all the schemas
I have edited decodes.properties:
EditDatabaseType=sql
EditDatabaseLocation=jdbc:postgresql://XXX
But it seems that I need to configure the DCS toolkit by running: launcher_start &
I cannot do that within my docker container.  NO GUI available here.
What decodes properties need to be set in order to configure opendcs to use postgresql?

It also seems that I need to generate a .decodes.auth file. It is not clear how to do this from the docs.
It would be much easier to just use environment variables, IMHO.

Thanks,
Pat.

**Comment by MikeNeilson** [comment, marked as answer]
```
jdbcDriverClass=org.postgresql.Driver
DbAuthFile=$DCSTOOL_USERDIR/.decodes.auth
editDatabseType=OPENTSDB
```

For the .decodes.auth file you can create that with
`setDecodesUser <filename>` and it'll prompt for the username and password. 

You should able to have that generated in the entry script on each run so it can be in environment files.

and yes I agree just plain put the dbhost /username/password in environment vars would easier. It's in the works. I'll see if I can finally get back to it next week, but in the mean time running the generation of the properties file and auth file from an entry script is a reasonable workaround.

FYI, You might consider setting this up in a docker-compose file, you and others would likely benefit from the ability to just say "docker-compose up -d" and have a working lrgs/DECODES/CP system. Would also save some money over starting AWS services just to test things.


Also for further configuration you *must* connect directly to the database. There are plans for an API but well, there's been some communications problems (using docker/docker-compose to start would avoid the need for a bastion host). (don't use the rest_api repo you see, that's got a major SQL injection total control vulnerability and the group that did it wants to just rewrite it.)

**Reply by cappelaere** [reply]
Thanks.  One problem with current approach is that the admin password is very vulnerable: simply run that command and pass the AuthFileName with a dash in front of it.  You will get the original username/password in plain text.

Yes.  Plan is to setup a docker-compose.  It will also include nginx and netdata.  The database is likely to be Aurora... but I am using AWS RDS right now for development. I would have liked to use TimestreamDB as a timeseries database.... Netdata will be used to gather metrics (Prometheus style likely... to be discussed later).  Thanks for your help.

**Reply by MikeNeilson** [reply]
Counter point: *only* if it's in the LRGSHOME folder, you can change that to a tmp directory that gets lost on restart.

Then your only concern would be attackers that can run docker on your live environment. not impossible, but much more difficult.

But let's consider this a prototype anyways, once you get it working at all and let us see the Dockerfile that'll be very helpful making the changes to pull from the environment directly.

**Comment by MikeNeilson** [comment]
Oh for clarity are you doing the "LRGS" database from the NOAA directory or the OPENTSDB-pg for DECODES?

**Reply by cappelaere** [reply]
Hummm.  Good question.  I loaded the schemas from the opendcs-pg folder.  I was under the impression that that database would contain the decoded messages and more (there is not much in the docs about this).  It would be great to clarify.
If I use the NOAA database, where are the messages stored?  Still in XML archives?

**Reply by MikeNeilson** [reply]
yeah, the LRGS message are only in the xml archive at the moment. The NOAA folder schema contains configuration elements for the LRGS configuration, but not the data. I'm looking into it at the moment. SQL wise won't be difficult, just adding tables. But converting the archive to an interface and adjusting may take a bit; need to review all of that code first before I start making changes.

**Reply by mikecovesw** [reply]
I'm not sure I'm following all of this discussion, but I thought I'd add 2 cents:

What you're calling "xml archive", I assume, are the day files (and corresponding index files) in the $LRGSHOME/archive directory, right? LRGS stores raw data only -- not decoded data. This is true whether you use a SQL DECODES database or not.

DECODES doesn't store decoded messages in the database either. Just the platform metadata required to translate a raw message into a decoded message in one of many output formats. There are "consumers" to inject time series data into CWMS, OpenTSDB (have you renamed that to something?), or HDB.

The schema in the noaa subdir is only used by NOAA Wallops, as far as I know. It contains the LRGS account info and usage stats. It does not contain raw message data.

There is schema in the opendcs-pg subdir used to store raw DCP transmissions. See the dcs_trans_* files. These are used by the DCP Monitor web application hosted at USACE MVR, NWP and a few other places. (I have one at covesw.com but the decoding info is not being actively maintained.)

Yes the passwords are visible from the setDecodesUser utility. But this can only be run if you login to the server and have permissions on file, which should be set to 600. I don't know how this plays with your "docker" stuff.

Finally, Pat: Most of the OpenDCS code was written in Ellicott City, MD. It's good to see you continue that tradition. Give my regards to Shannon's Saloon on Rt. 40!

**Reply by MikeNeilson** [reply]
Haven't renamed it yet, but we did establish the name will be "OpenHydroDB." I'll deal with the renaming in the 8.0 series branch; especially since that will be a fairly breaking change.

But yes, for "LRGS" the NOAA database still uses the xml daily files for the raw messages.

The DCPMonitor that Mike Mentioned is in another repository, that's fairly bog standard WAR. 

That said I do have plans to migrate everything to more interfaces and away from the XML files. SQLite, which I'm pretty sure either didn't exist or at least didn't have a java wrapper when DECODES was started, will cover the "i don't want server connections" use cases that exist in a few places.

**Reply by cappelaere** [reply]
@mikecovesw  Good to hear from you.  After 30 years in Ellicott City, I actually moved a little east to Pasadena, on the water...  Still not too far.  May be we could get together and have lunch or dinner if you still live in the area?
